### PR TITLE
Add LayerNorm support, add model splitting to `canonize`

### PR DESCRIPTION
**Pull request by Maximilian-Stefan-Ernst**
Add support for Flux `LayerNorm` and `Scale` layers.

`Scale` can be supported using the AD-fallback, but since there is no `.weight` field (it's called `.scale`), I wrote a method for `modify_layer` to make it compatible with the `GammaRule`.

`LayerNorm` is a bit more complicated, since it first normalizes its input, then applies a `Scale` layer. I decided to implement this with a custom `lrp!`-method that calls `lrp!` again for the `Scale` part. I'm not shure if this is the most reasonable way, and I'm open for suggestions.

For the implementation of the `normalise` part, i decided to use `Zygote.pullback`. This is because i realized that `LayerNorm(i, j, k, ...)` can be called with an arbitrary number of dimensions to normalize over, and this is quite annoying to implement directly. But it has the downside of adding `ChainRulesCore` as a dependency. Here I'm also open for suggestions.

Before the PR can be merged, there are some things to do/discuss:

- [x] add small epsilon in normalization layer to avoid devision by 0? (rules.jl, line 569)
- [x] add default rules for `LayerNorm`?
- [x] how to pass `keep_bias` in `lrp!`-method?
- [x] add documentation
    - [x] add paper on lrp for transformers to references
- [x] add tests
    - [x] check all edge cases for `LayerNorm` with and without affine layer and/or nonlinear activation
    - [x] do one lrp pass for a very small example by hand

### CODE REVIEW

File: src/rules.jl [hunk 1]
```diff
@@ -213,10 +213,22 @@ struct GammaRule{T<:Real} <: AbstractLRPRule
     γ::T
     GammaRule(gamma=LRP_DEFAULT_GAMMA) = new{eltype(gamma)}(gamma)
 end
+is_compatible(rule::GammaRule, layer::Scale) = true
 function modify_parameters(r::GammaRule, param::AbstractArray)
     γ = convert(eltype(param), r.γ)
     return @. param + γ * keep_positive(param)
 end
+function modify_layer(rule::GammaRule, layer::Scale; keep_bias=true)
```
Comment (adrhill) [inline src/rules.jl:221]: Do we have to specify the `GammaRule` here or can this be a more generic function compatible with other rules?

File: src/rules.jl [hunk 2]
```diff
@@ -518,6 +530,45 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+
+# Optional arguments
+- `affine_rule`: Rule applied to affine transformation, defaults to `ZeroRule()`.
+
+# References
+- 
+"""
+struct LayerNormRule{T <: AbstractLRPRule} <: AbstractLRPRule
+    affine_rule::T
+end
```
Comment (adrhill) [inline src/rules.jl:546]: Do we need this wrapper around other rules?
Maybe defining `lrp!(Rᵏ, rule, layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)` (without the limitation on `rule::LayerNormRule`) and using generic rules is enough?

File: src/rules.jl [hunk 3]
```diff
@@ -518,6 +530,45 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+    μ = mean(aᵏ, dims = dims)
+    σ = ignore_derivatives(std(aᵏ, dims=dims, mean=μ, corrected=false))
+    return @. (aᵏ - μ) / (σ + eps)
+end
+
+function lrp!(Rᵏ, rule::LayerNormRule, layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
+    # forward pass: split in normalization and scale
+    # aᵏ ->(normalize) aᵏₙ ->(scale) aᵏ⁺¹
+    ## normalize
+    aᵏₙ, back = Zygote.pullback(x -> detached_normalise(x, layer), aᵏ)
```
Comment (adrhill) [inline src/rules.jl:562]: I'd call this `aᵏ⁺¹ₙ` to indicate that it corresponds to a modified output of the layer.

File: src/rules.jl (hunk 2 above)
Comment (adrhill) [reply src/rules.jl:546]: RelevancePropagation is quite unique in its ability to assign rules to both layer and rule types, so a distinction between
```Julia
lrp!(Rᵏ, rule::LayerNormRule(GammaRule()), layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
```
and
```Julia
lrp!(Rᵏ, rule::GammaRule(), layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
```
might not be necessary.

**Comment by adrhill** [comment]
> add small epsilon in normalization layer to avoid devision by 0? (rules.jl, line 569)
 check all edge cases for LayerNorm with and without affine layer and/or nonlinear activation
 
See here: https://github.com/Julia-XAI/RelevancePropagation.jl/pull/9#discussion_r1483247721
 
> how to pass `keep_bias` in `lrp!`-method?

The `keep_bias` keyword-argument in `modify_layer` is currently not user-facing.
I was considering adding it as a field to all rules, but decided against it.
If we ever need this functionality, it should be added to all rules in another PR.
 
> add paper on lrp for transformers to references

Could you also put the equations in this PR for reference, so I can review the implementation?

File: src/rules.jl [hunk 4]
```diff
@@ -518,6 +530,42 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+    eps = convert(float(eltype(aᵏ)), layer.ϵ)
+    μ = mean(aᵏ; dims = 1:length(layer.size))
+    σ = std(aᵏ, dims=1:length(layer.size), mean=μ, corrected=false)
+    z = aᵏ - μ
+    aᵏₙ = @. z / (σ + eps)
+    # lrp pass
+    ## Rᵏ⁺¹ ->(scale) Rᵏₙ ->(normalize) Rᵏ
+    ## scale: call LRP on affine layer with "subrule" rule.affine_rule
+    Rᵏₙ = similar(aᵏₙ)
+    lrp!(Rᵏₙ, rule.affine_rule, layer.diag, modify_layer(rule.affine_rule, layer.diag), aᵏₙ, Rᵏ⁺¹)
```
Comment (adrhill) [inline src/rules.jl]: Just write into `Rᵏ` twice to save some allocations.
(This little trick should be documented.)

File: src/rules.jl [hunk 5]
```diff
@@ -518,6 +530,42 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+
+LayerNormRule() = LayerNormRule(ZeroRule())
+
+function lrp!(Rᵏ, rule::LayerNormRule, layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
+    # forward pass: split in normalization and scale
+    # aᵏ ->(normalize) aᵏₙ ->(scale) aᵏ⁺¹
+    ## normalize
+    eps = convert(float(eltype(aᵏ)), layer.ϵ)
+    μ = mean(aᵏ; dims = 1:length(layer.size))
+    σ = std(aᵏ, dims=1:length(layer.size), mean=μ, corrected=false)
```
Comment (adrhill) [inline src/rules.jl:542]: Can we introduce a variable for `1:length(layer.size)` since it gets used three times?

File: src/rules.jl (hunk 1 above)
Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:221]: Hmm, I think you know this better than me. `layer = Scale(n)` is basically a simpler `Dense` layer, that computes `x -> layer.scale .* x .+ layer.bias`.

Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:221]: So i guess `modify_weight` should work the same way on `layer.scale`; so it should also work for other rules.

Comment (adrhill) [reply src/rules.jl:221]: I'd remove the `GammaRule` restriction for now. Other weight modifications might be valid too.

File: src/rules.jl (hunk 4 above)
Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl]: done :)

File: src/rules.jl (hunk 5 above)
Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:542]: done :)

File: src/rules.jl (hunk 1 above)
Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:221]:
```suggestion
function modify_layer(rule, layer::Scale; keep_bias=true)
```

**Comment by Maximilian-Stefan-Ernst** [comment]
@adrhill: I'm not sure where I should document the behaviour of LRP for `LayerNorm` layers, since we don't have a rule for that anymore...

File: src/rules.jl [hunk 6]
```diff
@@ -518,6 +530,28 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+    n_dims = 1:length(layer.size)
+    eps = convert(float(eltype(aᵏ)), layer.ϵ)
+    μ = mean(aᵏ; dims=n_dims)
+    σ = std(aᵏ, dims=n_dims, mean=μ, corrected=false)
+    z = aᵏ - μ
+    aᵏₙ = @. z / (σ + eps)
+    # lrp pass
+    ## Rᵏ⁺¹ ->(scale) Rᵏ ->(normalize) Rᵏ
+    ## scale: call LRP on affine layer with "subrule" rule.affine_rule
+    lrp!(Rᵏ, rule, layer.diag, modify_layer(rule, layer.diag), aᵏₙ, Rᵏ⁺¹)
```
Comment (Maximilian-Stefan-Ernst) [inline src/rules.jl:549]: @adrhill: so it's okay for now that we can't pass `keep_bias` in this call to `modify_layer`?

Comment (adrhill) [reply src/rules.jl:549]: Yes!

**Comment by adrhill** [comment]
> I'm not sure where I should document the behaviour of LRP for LayerNorm layers, since we don't have a special rule anymore...

Yes, I agree this isn't great. In fact, this is a good argument for adding back an explicit `LayerNormRule`. Having no rule adds a lot of "magic" and there might be multiple rules for `LayerNorm` layers in the future.

While I'm not a big fan of rules wrapping other rules, I'll let you make the call on whether you want the functionality of calling a separate rule on the inner affine layer.

Sorry for the extra work! 🙇‍♂️

File: src/canonize.jl [hunk 7]
```diff
@@ -26,19 +30,68 @@ end
...
+
+    i = 1
+    while i <= length(model)
+        l = model[i]
+
+        if is_splittable(l)
+            splitted = split_layer(l)
+            model = Chain(model[1:(i - 1)]..., splitted..., model[(i + 1):end]...)
+            # if fused, don't increment i,
+            # instead try fusing the new layer with the next one
```
Comment (adrhill) [inline src/canonize.jl:79]: This comment still refers to `canonise_fuse`.

File: src/canonize.jl [hunk 8]
```diff
@@ -56,8 +109,19 @@ function _canonize(model::Chain)
...
+#= chainflatten(l::LayerNorm) = _chainflatten(l)
+function _chainflatten(l::LayerNorm)
+    if l.diag isa typeof(identity)
+        return l
+    else
+        norm = LayerNorm(identity, identity, l.ϵ, l.size, false)
+        diag = l.diag
+        return [norm, diag]
+    end
+end =#
```
Comment (adrhill) [inline src/canonize.jl:127]: This should be removed.

File: src/rules.jl [hunk 9]
```diff
@@ -533,7 +533,10 @@ end
 """
 #
 """
```
Comment (adrhill) [inline src/rules.jl:535]: This looks like it can be removed (?).

File: src/rules.jl [hunk 10]
```diff
@@ -546,10 +549,32 @@ function lrp!(Rᵏ, rule, layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
...
+
+function lrp!(Rᵏ, ::LayerNormRule, layer::LayerNorm{F,D}, _modified_layer, aᵏ, Rᵏ⁺¹) where {F,D<:typeof(identity)}
+    n_dims = 1:length(layer.size)
+    μₐ = mean(aᵏ; dims=n_dims)
+    s = @. Rᵏ⁺¹ / stabilize_denom(aᵏ - μₐ, LRP_DEFAULT_STABILIZER)
+    μₛ = mean(s; dims=n_dims)
+    @. Rᵏ = aᵏ * (s - μₛ)
+end
+
+function lrp!(Rᵏ, ::LayerNormRule, layer::LayerNorm, _modified_layer, aᵏ, Rᵏ⁺¹)
```
Comment (adrhill) [inline src/rules.jl:565]: It looks to me like this method has the same argument types as the one on L539.
I'm guessing the other one can be removed?

File: src/canonize.jl (hunk 7 above)
Comment (Maximilian-Stefan-Ernst) [reply src/canonize.jl:79]:
```suggestion
            # if splitted, don't increment i,
            # instead try splitting the first "sublayer" again
```

File: src/rules.jl (hunk 10 above)
Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:538]: Yes, sorry, I think the methods should now be correct.

File: src/rules.jl [hunk 11]
```diff
@@ -518,6 +519,43 @@ function lrp!(Rᵏ, rule::GeneralizedGammaRule, layer, modified_layers, aᵏ, R
...
+    lrp!(Rᵏ, ZeroRule(), scale, modify_layer(ZeroRule(), scale), aᵏₙ, Rᵏ⁺¹)
+    # compute LRP pass through normalization
+    s = @. Rᵏ / stabilize_denom(z, LRP_DEFAULT_STABILIZER)
+    μₛ = mean(s; dims=n_dims)
+    @. Rᵏ = aᵏ * (s - μₛ)
+end
+
+function lrp!(Rᵏ, ::ZeroRule, layer::LayerNorm, modified_layer, aᵏ, Rᵏ⁺¹)
+    lrp!(Rᵏ, LayerNormRule(), layer, modified_layer, aᵏ, Rᵏ⁺¹)
+end
```
Comment (adrhill) [inline src/rules.jl:557]: I think we should remove this as it silently switches the `ZeroRule` for the `LayerRule`:
* either we exclusively allow the `LayerNorm` rule on `LayerNorm` layers 
* or we also allow the `ZeroRule` AD fallback on L12

Comment (Maximilian-Stefan-Ernst) [reply src/rules.jl:573]: I don't have strong opinions on this, but I agree this needs to change, because my implementation does not allow to use the `ZeroRule` on `LayerNorm` at all, which is definitely a problem. Some things to consider:
1. Exclusively allow the `LayerNormRule`: this would make the default analyzer generated by `LRP` throw an error, because it assigns `ZeroRule` everywhere. So I guess we would need to change the default rule.
2. Allow the default fallback for `ZeroRule`: would be an option, but I think the "average" user maybe does not think too much about which rule to use for `LayerNorm`, and maybe giving them the more sensible default `LayerNormRule` would be nicer...

I think option 2 would be more sensible, but you should decide.

File: test/test_rules.jl [hunk 12]
```diff
@@ -247,6 +374,28 @@ layers = Dict(
...
+        @testset "$rulename" begin
+            for (layername, layer) in layers
+                @testset "$layername" begin
+                    run_rule_tests(rule, layer, rulename, layername, aᵏ_dense)
+                end
+            end
+        end
+    end
+end
+
```
Comment (Maximilian-Stefan-Ernst) [inline test/test_rules.jl:398]: I tried to do these tests analogous to the `Dense` tests but I did not really know what I was doing - maybe you can check if this is actually usefull :D

**Comment by Maximilian-Stefan-Ernst** [comment]
I think this is now (almost) finished. Some last things:
- [ ] The `LayerNormRule` should be added to the docs - maybe you could add them where you want them to be. I also added docstrings for the `LayerNormRule` but could not really check how they look because for some reasons, the docs take forever to build on my machine.
- [ ] I also added tests, but for some I'm not quite shure I did it right - I added a comment above.
- [ ] We need to resolve the above discussion on how `ZeroRule` should behave on `LayerNorm`

After that, I think you can run the CI and then merge.

**Comment by adrhill** [comment]
> for some reasons, the docs take forever to build on my machine.

You need to set the `DATADEPS_ALWAYS_ACCEPT=true` environment variable for MLDatasets.
This can either be done via the `ENV` dictionary within a Julia session of by calling 
```bash
DATADEPS_ALWAYS_ACCEPT=true julia --project=docs -e 'include("docs/make.jl")'
```

File: src/rules.jl (hunk 11 above)
Comment (adrhill) [reply src/rules.jl:573]: I'm in favor of option 2: 
The documentation and README already strictly discourage the use of the `ZeroRule` and LRP without composite. And I rather like that LRP with the `ZeroRule` corresponds to `InputTimesGradient`. This would be broken by using a custom `ZeroRule` implementation for LayerNorm.

**Comment by Maximilian-Stefan-Ernst** [comment]
Okay, I managed to build the docs locally - but on this branch it's still the old version. Should I merge main into this branch and add the documentation for `LayerNormRule` to the new docs version?

**Review by adrhill — APPROVED** [review]

**Comment by adrhill** [comment]
Since I just had to look this up, here is the derivation of our `LayerNorm` implementation (CC @chr5tphr):
___

Rule is defined as:

```math
R^k_i = \sum_j \frac{x_i \cdot(\delta_{ij} -\frac1N)}{\sum_{i'}x_{i'} \cdot(\delta_{i'j} -\frac1N)} R^{k+1}_j
```

with 

```math
\begin{align}
	z_j &= \sum_{i'}x_{i'} \cdot(\delta_{i'j} -\frac1N)\\
	&= x_j - \sum_{i'}\frac1N x_{i'}\\
	&= x_j - \mu_x
\end{align}
```

this simplifies to 

```math
\begin{align}
	R^k_i &= \sum_j \frac{x_i \cdot(\delta_{ij} -\frac1N)}{z_j} R^{k+1}_j\\
	&= x_i  \cdot \sum_j (\delta_{ij} -\frac1N) \underbrace{\frac{R^{k+1}_j}{z_j}}_{s_j}\\ 
	&= x_i \cdot \left( s_i - \sum_j \frac1N s_j \right)\\
	&= x_i \cdot \left( s_i - \mu_s \right)
\end{align}
```

or in Julia code (without batch dimension):
```julia
z = aᵏ .- mean(aᵏ) 
s = Rᵏ⁺¹ ./ z # use stabilize_denom 
Rᵏ .= aᵏ .* (s .- mean(s))
```

where in our notation $a^k = x$.
